File: frontend/mobile/modules/sentinel-collector/android/src/main/java/com/sentinel/collector/BufferPressure.kt
package com.sentinel.collector

/**
 * How the sample backlog is described where a person can see it.
 *
 * A bare "400 buffered" reads as healthy — a number going up looks like a
 * collector that is keeping its data. It is not: [SampleBuffer] is bounded, and
 * at the cap the oldest sample is *permanently discarded* on every new one.
 * Those two states look identical in a plain count and are completely different
 * facts, so the phrasing has to separate them.
 *
 * Found on a real phone: screen-off and idle for far longer than the buffer's
 * ~66 minutes of headroom, Doze cut network access despite the wake lock (a
 * limit [CollectorService] already documents the wake lock does not cover), the
 * buffer sat at exactly 400, history was being dropped — and the notification
 * said "400 buffered", which was read as fine.
 *
 * Loss is reported *after* it stops too. A backlog that drains looks like a
 * full recovery, and the samples that fell off the front are still gone; the
 * charts will have a hole in them that nothing else explains.
 *
 * Pure, so the wording is pinned by a JVM test. The React Native side phrases
 * the same three facts for the collector screen — it cannot import this, so
 * both surfaces render `buffered`, `capacity` and `dropped` rather than one of
 * them inventing a number the other does not have.
 */
object BufferPressure {

    /**
     * At or above this fraction of capacity, say how close the cap is instead
     * of only how much is waiting. The warning is worth having before data is
     * lost, not only once it already has been.
     */
    const val NEAR_FULL_FRACTION = 0.9

    /** The backlog phrase for the persistent notification, or null when there
     *  is nothing waiting and nothing has been lost this run. */
    fun describe(buffered: Int, capacity: Int, dropped: Int): String? {
        val lost = if (dropped > 0) " ($dropped lost)" else ""
        return when {
            capacity > 0 && buffered >= capacity ->
                "$buffered buffered — full, dropping oldest$lost"
            capacity > 0 && buffered >= capacity * NEAR_FULL_FRACTION ->
                "$buffered of $capacity buffered — nearly full$lost"
            buffered > 0 -> "$buffered buffered$lost"
            dropped > 0 -> "$dropped sample${plural(dropped)} lost while the buffer was full"
            else -> null
        }
    }

    private fun plural(count: Int) = if (count == 1) "" else "s"
}

File: frontend/mobile/src/lib/bufferPressure.ts
/**
 * How the collector's sample backlog reads on the collector screen.
 *
 * The TypeScript half of the collector module's `BufferPressure.kt`, which does
 * the same job for the persistent notification. Neither can import the other,
 * so the rule is that both render the same three *facts* the service reports —
 * buffered, capacity, dropped — and neither invents a number the other does not
 * have.
 *
 * What this exists to prevent: the ring buffer is bounded (400 samples, about
 * 66 minutes at the 10s cadence) and at the cap it discards the oldest sample
 * for every new one. That is the right behaviour, but it is *loss*, not delay,
 * and a plain count cannot tell the two apart — a phone left idle overnight had
 * been dropping history for hours behind a number that read as healthy.
 *
 * Pure, so the wording is pinned by a test rather than by whichever screen
 * happened to render it, the same reason `chartFrame.ts` and `deviceReadings.ts`
 * live here instead of inside their components.
 */

export interface BufferPressure {
  /** The row value, always something — this row is not conditionally rendered. */
  text: string
  /** True once samples have actually been discarded: the screen calls this out
   *  rather than leaving it as one grey row among several. */
  losing: boolean
}

/** At or above this fraction of capacity, name the cap. The warning is worth
 *  having while there is still something to be done about it. */
export const NEAR_FULL_FRACTION = 0.9

export function describeBuffer(status: {
  bufferedSamples: number
  bufferCapacity: number
  droppedSamples: number
}): BufferPressure {
  const { bufferedSamples: buffered, bufferCapacity: capacity, droppedSamples: dropped } = status
  const lost = dropped > 0 ? ` · ${dropped} lost` : ""

  if (capacity > 0 && buffered >= capacity) {
    return { text: `${buffered} — full, dropping oldest${lost}`, losing: true }
  }
  if (dropped > 0) {
    // The backlog has drained, or is draining, and the hole in the history is
    // still there. Reporting only the current size here would read as a full
    // recovery.
    return {
      text: buffered === 0 ? `${dropped} lost while full` : `${buffered}${lost}`,
      losing: true,
    }
  }
  if (capacity > 0 && buffered >= capacity * NEAR_FULL_FRACTION) {
    return { text: `${buffered} of ${capacity} — nearly full`, losing: false }
  }
  if (buffered === 0) {
    return { text: "nothing waiting", losing: false }
  }
  return { text: `${buffered} sample${buffered === 1 ? "" : "s"}`, losing: false }
}
